src/fsharp/Optimizer.fs

```diff
@@ -2543,6 +2543,37 @@ and TryDevirtualizeApplication cenv env (f,tyargs,args,m) =
                 MightMakeCriticalTailcall = false;
                 Info=UnknownValue})
 
+    // Analyze the name of the given symbol and rewrite AST to constant string expression with the name
+    | Expr.Val(vref,_,_),_,_ when valRefEq cenv.g vref cenv.g.nameof_vref ->
```

Comment:
Please check this works on a generic value

    let f x = x
    nameof(f)

I don't believe it will, since a use of a generic value has a type application wrapped around it, indeed it is likely to create invalid code.